internal/cephfs/controllerserver.go

```diff
@@ -423,3 +428,308 @@ func (cs *ControllerServer) ControllerExpandVolume(ctx context.Context, req *csi
+					requestName, errDefer)
+			}
+		}
+	}()
+	snap := snapshotInfo{}
+	snap, err = doSnapshot(ctx, parentVolOptions, vid.FsSubvolName, sID.FsSnapshotName, cr)
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	return &csi.CreateSnapshotResponse{
+		Snapshot: &csi.Snapshot{
+			SizeBytes:      int64(info.BytesQuota),
+			SnapshotId:     sID.SnapshotID,
+			SourceVolumeId: req.GetSourceVolumeId(),
+			CreationTime:   snap.CreationTime,
+			ReadyToUse:     true,
+		},
+	}, nil
+}
+
+func doSnapshot(ctx context.Context, volOpt *volumeOptions, subvolumeName, snapshotName string, cr *util.Credentials) (snapshotInfo, error) {
+	volID := volumeID(subvolumeName)
+	snapID := volumeID(snapshotName)
+	snap := snapshotInfo{}
+	err := createSnapshot(ctx, volOpt, cr, snapID, volID)
+	if err != nil {
+		klog.Errorf(util.Log(ctx, "failed to create snapshot %s %v"), snapID, err)
+		return snap, err
+	}
+	defer func() {
+		if err != nil {
+			dErr := deleteSnapshot(ctx, volOpt, cr, snapID, volID)
+			if dErr != nil {
+				klog.Errorf(util.Log(ctx, "failed to delete snapshot %s %v"), snapID, err)
+			}
+		}
+	}()
+	snap, err = getSnapshotInfo(ctx, volOpt, cr, snapID, volID)
+	if err != nil {
+		klog.Errorf(util.Log(ctx, "failed to get snapshot info %s %v"), snapID, err)
+		return snap, fmt.Errorf("failed to get snapshot info for snapshot:%s", snapID)
+	}
+	var t *timestamp.Timestamp
+	t, err = parseTime(ctx, snap.CreatedAt)
+	if err != nil {
+		return snap, err
+	}
+	snap.CreationTime = t
+	err = protectSnapshot(ctx, volOpt, cr, snapID, volID)
+	if err != nil {
+		klog.Errorf(util.Log(ctx, "failed to protect snapshot %s %v"), snapID, err)
+	}
+	return snap, err
+}
+
+func parseTime(ctx context.Context, createTime string) (*timestamp.Timestamp, error) {
+	tm := &timestamp.Timestamp{}
+	layout := "2006-01-02 15:04:05.000000"
+	// TODO currently parsing of timestamp to time.ANSIC generate from ceph fs is failng
+	var t time.Time
+	t, err := time.Parse(layout, createTime)
+	if err != nil {
+		klog.Errorf(util.Log(ctx, "failed to parse time %s %v"), createTime, err)
+		return tm, err
+	}
+	tm, err = ptypes.TimestampProto(t)
+	if err != nil {
+		klog.Errorf(util.Log(ctx, "failed to convert time %s %v"), createTime, err)
+		return tm, err
+	}
+	return tm, nil
+}
+func (cs *ControllerServer) validateSnapshotReq(ctx context.Context, req *csi.CreateSnapshotRequest) error {
+	if err := cs.Driver.ValidateControllerServiceRequest(csi.ControllerServiceCapability_RPC_CREATE_DELETE_SNAPSHOT); err != nil {
+		klog.Errorf(util.Log(ctx, "invalid create snapshot req: %v"), protosanitizer.StripSecrets(req))
+		return err
+	}
+
+	// Check sanity of request Snapshot Name, Source Volume Id
+	if req.Name == "" {
+		return status.Error(codes.InvalidArgument, "snapshot Name cannot be empty")
+	}
+	if req.SourceVolumeId == "" {
+		return status.Error(codes.InvalidArgument, "source Volume ID cannot be empty")
```

Comment:
is this correct, or should it be `codes.NotFound`?

Reply:
we can return NotFound and changed to NotFound now.